src/wp-includes/blocks.php

```diff
@@ -2713,18 +2714,59 @@ function build_query_vars_from_query_block( $block, $page ) {
 			}
 			$query['tax_query'] = array_merge( $query['tax_query'], $tax_query_back_compat );
 		}
-		if ( ! empty( $block->context['query']['taxQuery'] ) ) {
-			$tax_query = array();
-			foreach ( $block->context['query']['taxQuery'] as $taxonomy => $terms ) {
-				if ( is_taxonomy_viewable( $taxonomy ) && ! empty( $terms ) ) {
-					$tax_query[] = array(
-						'taxonomy'         => $taxonomy,
-						'terms'            => array_filter( array_map( 'intval', $terms ) ),
-						'include_children' => false,
-					);
+
+		if ( ! empty( $block->context['query']['taxQuery'] ) && is_array( $block->context['query']['taxQuery'] ) ) {
+			$tax_query_input = $block->context['query']['taxQuery'];
+			$tax_query       = array();
+			// If there are keys other than include/exclude, it's the old
+			// format e.g. "taxQuery":{"category":[4]}
+			if ( ! empty( array_diff( array_keys( $tax_query_input ), array( 'include', 'exclude' ) ) ) ) {
+				foreach ( $block->context['query']['taxQuery'] as $taxonomy => $terms ) {
+					if ( is_taxonomy_viewable( $taxonomy ) && ! empty( $terms ) ) {
+						$tax_query[] = array(
+							'taxonomy'         => $taxonomy,
+							'terms'            => array_filter( array_map( 'intval', $terms ) ),
+							'include_children' => false,
+						);
+					}
 				}
+			} else {
+				// This is the new format e.g. "taxQuery":{"include":{"category":[4]},"exclude":{"post_tag":[5]}}
+
+				// Helper function to build tax_query conditions from taxonomy terms.
+				$build_conditions = static function ( array $terms, string $operator = 'IN' ): array {
+					$conditions = array();
```

Comment:
I'm partially going back on myself here. But if there were block attributes which were seriously corrupted, like `"taxQuery":{"include":{"category":123},"exclude":{"post_tag":null}}` then the following would cause a fatal error. So this would make it more resilient:

```suggestion
				$build_conditions = static function ( $terms, string $operator = 'IN' ): array {
					if ( ! is_array( $terms ) ) {
						return array();
					}
					$conditions = array();
```

Reply:
Alternatively, it could do:


```suggestion
				$build_conditions = static function ( $terms, string $operator = 'IN' ): array {
					$terms      = (array) $terms;
					$conditions = array();
```